# aspergillusAculeatus
# Predictive model of cellulase gene expression in A. aculeatus 
To quantitatively investigate the genes or genomic features that highly “influence” the enzyme activities, we first performed multiple linear regressions using the TPM values from the transcriptomes described earlier as the “predictor” variables, and the enzyme activities of the matched strains and samples as the “response” variables. We focused on the TPM of 8,623 genes shared by the three strains, which were taken from at least 3 independent replicates of each of the three strains.s. Firstly, the trend of data was explored and illustrated by scatter plots using the package “pandas” and “seaborn” in Python. Within the genes shared by all the three strains, 1,125 “important” features, which were genes/transcripts in this case, were pre-selected using the Information gain (GI) criteria based on the weight above 0.95 given by each feature selection (Figure 5). All selected features (Appendix Table 1) were then used to construct a multiple linear regression model by function “LinearRegression” in the package “Scikit-learn”, also in Python?. Finally, the performance of our model was assessed in terms of Ajd. R-squared, F-statistic, and coefficient values using the Python package “statsmodels”. All steps in model construction were performed under the Jupyter Notebook, an interactive application for editing and running python codes via a web browser.
